src/osd/PG.cc

```diff
@@ -1561,9 +1561,14 @@ void PG::choose_async_recovery_ec(const map<pg_shard_t, pg_info_t> &all_info,
     // past the authoritative last_update the same as those equal to it.
     version_t auth_version = auth_info.last_update.version;
     version_t candidate_version = shard_info.last_update.version;
-    if (auth_version > candidate_version &&
-        (auth_version - candidate_version) > cct->_conf.get_val<uint64_t>("osd_async_recovery_min_pg_log_entries")) {
-      candidates_by_cost.insert(make_pair(auth_version - candidate_version, shard_i));
+    uint64_t approx_missing_objects =
+      shard_info.stats.stats.sum.num_objects_missing;
+    if (auth_version > candidate_version) {
+      approx_missing_objects += auth_version - candidate_version;
+    }
+    if (approx_missing_objects > cct->_conf.get_val<uint64_t>(
```

Comment:
@xiexingguo If `num_objects_missing` is reliable, this change should work. Wondering if you have done some evaluation like https://github.com/ceph/ceph/pull/23663#issuecomment-416898500, to compare how this change impacts the overall performance of async recovery.

Since this change is critical to how async recovery works in general, I'd like @jdurgin to review this as well.